drivers/hmis/app/models/hmis/ce/README_FOR_CHANGE_MARKER.md

```diff
@@ -0,0 +1,98 @@
+# Coordinated Entry (CE) Change Tracking System
+
+The change tracking system for Coordinated Entry (CE) efficiently processes client and housing opportunity updates for eligibility and prioritization.
+
+## System Overview
+
+The change tracking system addresses the performance challenges of full reprocessing by implementing an incremental update mechanism. Instead of re-evaluating all clients and candidate pools whenever data changes, this system identifies only the records that have been modified ("dirty" records) and processes them in batches.
+
+This is achieved through a versioning system managed by the `Hmis::Ce::ChangeMarker` model and a self-scheduling background job, `Hmis::Ce::ProcessChangesJob`, which continuously processes these changes.
+
+### Core Components
+
+- **`Hmis::Ce::ChangeMarker`**: A polymorphic model that tracks the state of other records (currently `GrdaWarehouse::Hud::Client` and `Hmis::Ce::Match::CandidatePool`). It uses `current_version` and `processed_version` to determine if a record is "dirty."
+- **`Hmis::Ce::ProcessChangesJob`**: A self-scheduling job that runs continuously to process dirty records. It uses an advisory lock to prevent concurrent runs.
+- **`Hmis::Ce::BuildCandidatePoolsJob`**: A job triggered by actions like marking units available. It creates or updates candidate pools and marks them as dirty for processing.
+- **`Hmis::MarkClientAsDirtyBehavior`**: A concern included in various HUD models to automatically mark a client as dirty whenever their data is saved.
+
+## Workflow
+
+The following diagram illustrates the flow of data from a user action to final processing:
+
+```mermaid
+sequenceDiagram
+    participant UserAction as User Action<br/>(e.g., Edit Client)
+    participant RailsApp as Rails Application
+    participant MarkClientAsDirtyBehavior as MarkClientAsDirtyBehavior
+    participant IdentifyDuplicates as IdentifyDuplicates
+    participant ChangeMarker as Hmis::Ce::ChangeMarker
+    participant BuildCandidatePoolsJob as Hmis::Ce::BuildCandidatePoolsJob
+    participant ProcessChangesJob as Hmis::Ce::ProcessChangesJob
+
+    UserAction->>+RailsApp: Submits a change (e.g., assessment)
+    RailsApp->>+MarkClientAsDirtyBehavior: after_save callback
+    MarkClientAsDirtyBehavior->>+ChangeMarker: upsert_or_bump_version (source client)
+    ChangeMarker-->>-MarkClientAsDirtyBehavior: Increments `current_version`
+    MarkClientAsDirtyBehavior-->>-RailsApp: Done
+
+    Note over RailsApp,IdentifyDuplicates: Client creation/update triggers deduplication
+    RailsApp->>+IdentifyDuplicates: Runs periodically or on client changes
+    IdentifyDuplicates->>+ChangeMarker: upsert_or_bump_version (destination client)
+    ChangeMarker-->>-IdentifyDuplicates: Marks destination client dirty
+    IdentifyDuplicates-->>-RailsApp: Done
+
+    alt Client or Pool Changes
+        UserAction->>+RailsApp: Mark Units Available
```

Comment:
There is another "Dev User Action" that requires reprocessing pools: making changes to CE Match Rules. Right now I am making Rule changes manually in the console, and then my process has been to run `HmisUtil::CeBuilder.build_candidate_pools` to rebuild the candidate pools. Is that still the right process to follow, after this change?

Reply:
updated note to self: can run `Hmis::Ce::BuildCandidatePoolsJob.perform_now` instead after changing match rules. worker must be running in order to pick up the dirty marker and actually reprocess the pool(s)

Reply:
I adjusted HmisUtil::CeBuilder.build_candidate_pools so that it calls the appropriate jobs, simulating what should happen async.